x/ibc/07-tendermint/types/header.go

```diff
@@ -15,23 +15,31 @@ import (
 var _ clientexported.Header = Header{}
 
 // Header defines the Tendermint consensus Header
+// Header encapsulates all the information necessary to update from a trusted Tendermint ConsensusState
+// The SignedHeader and ValidatorSet are the new untrusted update fields for the client
+// The TrustedHeight is the height of a stored ConsensusState on the client that will be used to verify new untrusted header
+// The Trusted ConsensusState must be within the unbonding period of current time in order to correctly verify,
+// and the TrustedValidators must hash to TrustedConsensusState.NextValidatorsHash since that is the last trusted validatorset
+// at the TrustedHeight
 type Header struct {
 	tmtypes.SignedHeader `json:"signed_header" yaml:"signed_header"` // contains the commitment root
-	ValidatorSet         *tmtypes.ValidatorSet                       `json:"validator_set" yaml:"validator_set"`
+	ValidatorSet         *tmtypes.ValidatorSet                       `json:"validator_set" yaml:"validator_set"`   // the validator set that signed Header
+	TrustedHeight        uint64                                      `json:"trusted_height" yaml:"trusted_height"` // the height of a trusted header seen by client less than or equal to Header
+	TrustedValidators    *tmtypes.ValidatorSet                       `json:"trusted_vals" yaml:"trusted_vals"`     // the last trusted validator set at trusted height
```

Comment:
Note that the new Header contains both the `TrustedValidators` AND the `TrustedHeight`.

This allows us to remove the ValidatorSet from the ConsensusState, and it also removes race conditions when relayers when they try to update a client.

Suppose a client is at height `h`, and a relayer1 decides to update to `h+3` with `TrustedHeight = h` and a relayer2 decides to update to `h+5` with `TrustedHeight=h`.

Both update msgs will be successful regardless of processing order. This was not true before, and would not have been true if we implicity assumed TrustedHeight was greatest stored height less than header height

Reply:
this is also worth documenting it